src/v/storage/segment_index.cc

```diff
@@ -177,16 +203,18 @@ ss::future<bool> segment_index::materialize_index() {
 
 ss::future<> segment_index::drop_all_data() {
     reset();
-    return _out.truncate(0);
+    co_return co_await (co_await open()).truncate(0);
 }
 
 ss::future<> segment_index::flush() {
     if (!_needs_persistence) {
         co_return;
     }
     _needs_persistence = false;
-    co_await _out.truncate(0);
-    auto out = co_await ss::make_file_output_stream(ss::file(_out.dup()));
+    auto backing_file = co_await open();
+    co_await backing_file.truncate(0);
+    auto out = co_await ss::make_file_output_stream(
+      ss::file(backing_file.dup()));
```

Comment:
The documentation for `ss::file::dup` says "Use on read-only files." but this is not a read-only file?